Fix thread leak in worker API

ghale opened the pull request:
This fixes an issue where worker API invocations could eagerly expand the thread pool way beyond what's necessary to process the work submitted.  This PR does two things:

1. Only expand the thread pool when there is incomplete work associated with this build operation (for instance, when nested work has been submitted) and there is work waiting to be processed.
1. Moves the work queue up to build session scope (from project scope) so that all projects use the same queue.

oehme commented on subprojects/core/src/main/java/org/gradle/internal/work/DefaultAsyncWorkTracker.java:
```diff
@@ -92,6 +92,22 @@ public void run() {
+    public boolean hasUncompletedWork(BuildOperationRef operation) {
+        lock.lock();
+        try {
+            List<AsyncWorkCompletion> workItems = ImmutableList.copyOf(items.get(operation));
```
⭕️ Why the copy?

ghale replied:
Bleh, cut/paste error.  I've removed the copy.

oehme commented on subprojects/core/src/main/java/org/gradle/internal/work/DefaultAsyncWorkTracker.java:
```diff
@@ -92,6 +92,22 @@ public void run() {
+        lock.lock();
+        try {
+            List<AsyncWorkCompletion> workItems = ImmutableList.copyOf(items.get(operation));
+            return CollectionUtils.any(workItems, new Spec<AsyncWorkCompletion>() {
```
I personally prefer a simple for-loop below Java 8. It just creates less noise.

ghale replied:
Sure, why not.  Done.

oehme commented on subprojects/base-services/src/main/java/org/gradle/internal/work/DefaultConditionalExecutionQueue.java:
```diff
@@ -78,10 +78,16 @@ public void submit(ConditionalExecution<T> execution) {
+        expand(false);
+    }
+
+    private void expand(boolean force) {
```
⭕️ This is kinda hard to understand and could use some (class/method-level) documentation

ghale replied:
Yeah, good point.  I've added some comments to (hopefully) make the function of this thing a little clearer.

big-guy commented:
+1 on some comments, lgtm

ghale commented:
Thanks for the feedback, guys.  I've addressed the comments and I also added some test coverage for checking the thread pool size when running nested operations as well (with a few minor functionality changes).  Unless you have further comments, I'll plan on merging once the latest changes pass CI.
